src/frequenz/sdk/timeseries/logical_meter/_formula_engine.py

```diff
@@ -0,0 +1,454 @@
+    ) -> datetime:
+        """Synchronize the metric streams.
+
+        For synchronised streams like data from the `ComponentMetricsResamplingActor`,
+        this a call to this function is required only once, before the first set of
+        inputs are fetched.
+
+        Args:
+            metrics: The finished tasks from the first `fetch_next` calls to all the
+                `MetricFetcher`s.
+
+        Returns:
+            The timestamp of the latest metric value.
+
+        Raises:
+            ValueError: when some streams have no value, or when the synchronization of
+                timestamps fails.
+        """
+        metrics_by_ts: Dict[datetime, str] = {}
+        for metric in metrics:
+            result = metric.result()
+            name = metric.get_name()
+            if result is None:
+                raise ValueError(f"Stream closed for component: {name}")
+            metrics_by_ts[result.timestamp] = name
+        latest_ts = max(metrics_by_ts)
+
+        # fetch the metrics with non-latest timestamps again until we have the values
+        # for the same ts for all metrics.
+        for metric_ts, name in metrics_by_ts.items():
+            if metric_ts == latest_ts:
+                continue
+            fetcher = self._metric_fetchers[name]
+            while metric_ts < latest_ts:
+                next_val = await fetcher.fetch_next()
+                assert next_val is not None
+                metric_ts = next_val.timestamp
+            if metric_ts > latest_ts:
+                raise ValueError(
+                    "Unable to synchronize timestamps of resampled metrics"
+                )
+        self._first_run = False
+        return latest_ts
+
+    async def apply(self) -> Sample:
+        """Fetch the latest metrics, apply the formula once and return the result.
+
+        Returns:
+            The result of the formula.
+
+        Raises:
+            ValueError: if some samples didn't arrive, or if formula application failed.
+        """
+        eval_stack: List[float] = []
+        ready_metrics, pending = await asyncio.wait(
+            [
+                asyncio.create_task(fetcher.fetch_next(), name=name)
+                for name, fetcher in self._metric_fetchers.items()
+            ],
+            return_when=asyncio.ALL_COMPLETED,
+        )
+
+        if pending or any(res.result() is None for res in iter(ready_metrics)):
+            raise ValueError("Some resampled metrics didn't arrive")
+
+        if self._first_run:
+            metric_ts = await self.synchronize_metric_timestamps(ready_metrics)
+        else:
+            res = next(iter(ready_metrics)).result()
+            assert res is not None
+            metric_ts = res.timestamp
+
+        for step in self._steps:
+            step.apply(eval_stack)
+
+        if len(eval_stack) != 1:
+            raise ValueError("Formula application failed.")
+
+        return Sample(metric_ts, eval_stack[0])
+
+
+class FormulaBuilder:
```

Comment:
Even though it's small, I think that this class deserves it's own file.

Reply:
done